htmd/parameterization/dihedral.py

```diff
@@ -330,41 +345,91 @@ def _evaluateConstTerms(self):
+
+    def _fit(self):
         # Save the initial parameters
         vector = self._paramsToVector(self.parameters)
 
         # Evaluate constant terms
-        self._const_energies = self._evaluateConstTerms()
+        const_energies = self._evaluateConstTerms()
 
         # Evaluate target energies for fitting
-        self._target_energies = []
-        for ref_energies, const_energies in zip(self._reference_energies, self._const_energies):
-            energies = ref_energies - const_energies
-            self._target_energies.append(energies)
-
-        shift = np.min(np.concatenate(self._target_energies))
-        self._target_energies = [energies - shift for energies in self._target_energies]
+        target_energies = self._calculateTargetEnergies(const_energies)
 
         # Check self-consistency of computed energies
-        self._objective(vector)
-        test_energies = zip(self._initial_energies, self._const_energies, self._actual_energies)
+        k0, phi0, offset, n = self._unpackVector(vector)
+        actual_energies = _evaluateScanEnergies(k0, phi0, offset, n, self._angle_values, self.numDihedrals, np.arange(self.numDihedrals))
+        test_energies = zip(self._initial_energies, const_energies, actual_energies)
         for initial_energies, const_energies, actual_enegies in test_energies:
             assert np.allclose(initial_energies, const_energies + actual_enegies, rtol=0, atol=1e-5) # TODO debug
 
         # Zero the initial parameters, so they are not used to start the parameter fitting
         if self.zeroed_parameters:
             vector[:] = 0
 
-        # Optimize the parameters
-        logger.info('Start parameter optimization')
+        # Optimize each dihedral individually
+        logger.info('Start parameter optimization with {} iterations'.format(self.num_iterations))
         start = time.clock()
-        vector = self._optimizeWithRandomSearch(vector)
+        best_loss = None
+        best_vector = None
+        for i in range(self.num_iterations):
+            logger.info('Dihedral fitting iteration {}'.format(i+1))
+
+            # Evaluate constant terms for single dihedral fitting
+            current_params = self._vectorToParams(vector)
+            const_energies_dihedral = self._evaluateConstTermsPerDihedral(current_params)
+            target_energies_dihedral = self._calculateTargetEnergies(const_energies_dihedral)
+
+            # Fit the individual dihedrals
+            dihedral_vectors = []
+            for idihed in range(self.numDihedrals):
+                logger.info('Optimizing dihedral {}/{}'.format(idihed+1, self.numDihedrals))
+                subvector = self._paramsToVector(current_params, idihed)
+                subvector = self._optimizeWithRandomSearch(subvector, target_energies_dihedral, num_searches=20, idihed=idihed, _logger=False)
+                dihedral_vectors.append(subvector)
+
+            # Pack the individually fitted parameters into a vector
```

Comment:
This could be factored out to `self._subvectorToVector` or something like that.

Reply:
Done